# SNaQ Tutorial: from alignments to phylogenetic network

This website has an alternative tutorial for SNaQ to the one in the [PhyloNetworks wiki](https://github.com/crsl4/PhyloNetworks.jl/wiki). Here, we use a Docker for the installation of Dependencies and we use a real dataset (baobabs) to illustrate the pipeline from multilocus alignments to phylogenetic network.

This learning materials were created as part of the workshop in the Kew Royal Botanical Gardens: [Methodological Advances in Reticulate Evolution](https://gtiley.github.io/RBG-Networks/about/) taught in November 2023 co-organized with George Tiley, Ziheng Yang and Tomas Flouri.


## Learning outcomes

By the end of the tutorial, you will be able to
- estimate gene trees (and posterior distributions) from multilocus alignments with [MrBayes](http://nbisweden.github.io/MrBayes/)
- estimate concordance factors from the estimated posterior distributions with [BUCKy](http://pages.stat.wisc.edu/~ane/bucky/index.html)
- estimate a phylogenetic network from the estimated concordance factors with [SNaQ](https://github.com/crsl4/PhyloNetworks.jl)


## Schedule

| Topic | Lecture notes | Time |
| :---:   | :---: | :---:       |
| 1. Theory: Estimating phylogenetic networks from multilocus alignments | [notes](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/lecture1.html) | 13:00 - 15:00 |
| Break | | 15:00 - 15:30 |
| 2. Set up: dependencies and data | [set-up](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/set-up.html), [data](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/lecture2.html)  | 15:30 - 16:30  |
| 3. From alignments to concordance factors with the TICR pipeline |  [notes](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/lecture3.html) | 16:30 - 17:30 |
| Q&A | | 15:30 - 16:00 |

| Topic | Lecture notes | Time |
| :---:   | :---: | :---:       |
| 4. Estimating a phylogenetic network with SNaQ | [notes](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/lecture4.html) | 9:00 - 10:30 |
| Break | | 10:30 - 11:00 |
| 5. Comparative models for trait evolution on networks | [notes](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/lecture5.html) | 11:00 - 12:00 |

## Citations

If you use the TICR pipeline, PhyloNetworks and/or SNaQ for your work, please cite the references in [here](https://solislemuslab.github.io/snaq-tutorial//lecture-notes/citations.html)

## Meet the instructor

Claudia Sol&iacute;s-Lemus is an assistant professor at the [Wisconsin Institute for Discovery](https://wid.wisc.edu/) and the [Department of Plant Pathology](https://plantpath.wisc.edu/) at the [University of Wisconsin-Madison](http://www.wisc.edu). Originally from Mexico City, she did her Undergraduate degrees
in Actuarial Sciences and Applied Mathematics at [ITAM](https://www.itam.mx/en).
Then, she did a MA in [Mathematics](http://www.math.wisc.edu) and a PhD in [Statistics](http://www.stat.wisc.edu) at the [University of Wisconsin-Madison](http://www.wisc.edu). 

<div class="container">
    <div class="row">
        <div class="column">
            <a href="https://solislemuslab.github.io/snaq-tutorial/images/claudiaSmall1.png">
            <img src="https://solislemuslab.github.io/snaq-tutorial/images/claudiaSmall1.png" width="150"
                  title="Claudia Sol&iacute;s-Lemus" alt="Claudia Sol&iacute;s.Lemus"/></a>
        </div>
        <div class="column">
            Pronouns: she/her <br/>
            <a href="https://namedrop.io/claudiasolislemus">Name pronunciation</a><br/>
            <a href="https://solislemuslab.github.io/">Lab website</a><br/>
            <a href="https://scholar.google.com/citations?user=GrUypj8AAAAJ&hl=en&oi=ao">Google scholar</a><br/>
            <a href="https://github.com/crsl4">GitHub</a><br/>
            <a href="https://solislemuslab.github.io//pages/people.html">Contact Info</a><br/>
        </div>
    </div>
</div>


